lib/segment/src/vector_storage/query/reco_query.rs

```diff
@@ -87,33 +88,101 @@ impl From<RecoQuery<VectorType>> for RecoQuery<Vector> {
+        #[case] chosen: Chosen,
         #[case] expected: ScoreType,
     ) {
         let query = RecoQuery::new(positives, negatives);
 
         let dummy_similarity = |x: &isize| *x as ScoreType;
 
+        let positive_transformation = scaled_fast_sigmoid;
+        let negative_transformation = |x| -scaled_fast_sigmoid(x);
+
         let score = query.score_by(dummy_similarity);
 
-        assert_eq!(score, expected);
+        match chosen {
+            Chosen::Positive => {
+                assert_eq!(score, positive_transformation(expected));
+            }
+            Chosen::Negative => {
+                assert_eq!(score, negative_transformation(expected));
+            }
+        }
+    }
+
+    proptest! {
+        /// Checks that the negative-chosen scores invert the order of the candidates
+        #[test]
+        fn correct_negative_order(a in -100f32..=100f32, b in -100f32..=100f32) {
+            let dummy_similarity = |x: &f32| *x as ScoreType;
+
+            let positive_ordering = dummy_similarity(&a).total_cmp(&dummy_similarity(&b));
+
+            let query_a = RecoQuery::new(vec![], vec![a]);
+            let query_b = RecoQuery::new(vec![], vec![b]);
+
+            let negative_ordering = query_a.score_by(dummy_similarity).total_cmp(&query_b.score_by(dummy_similarity));
+
+            if positive_ordering == std::cmp::Ordering::Equal {
+                assert_eq!(positive_ordering, negative_ordering);
+            } else {
+                assert!(positive_ordering != negative_ordering)
+            }
+        }
+
+        /// Checks that the positive-chosen scores preserve the order of the candidates
+        #[test]
+        fn correct_positive_order(a in -100f32..=100f32, b in -100f32..=100f32) {
+            let dummy_similarity = |x: &f32| *x as ScoreType;
+
+            let positive_ordering = dummy_similarity(&a).total_cmp(&dummy_similarity(&b));
+
+            let query_a = RecoQuery::new(vec![a], vec![]);
+            let query_b = RecoQuery::new(vec![b], vec![]);
+
+            let positive_ordering_after = query_a.score_by(dummy_similarity).total_cmp(&query_b.score_by(dummy_similarity));
+
+            assert_eq!(positive_ordering, positive_ordering_after);
+        }
+
+        /// Guarantees that if the positive and negative chosen points yield within (-50, 50)
+        /// similarities, the point that was chosen from positive is always preferred on
+        /// the candidate list
+        #[test]
+        fn correct_positive_and_negative_order(p in -100f32..=100f32, n in -100f32..=100f32) {
+            let dummy_similarity = |x: &f32| *x as ScoreType;
+
+            let query_p = RecoQuery::new(vec![p], vec![]);
+            let query_n = RecoQuery::new(vec![], vec![n]);
+
+            let ordering = query_p.score_by(dummy_similarity).total_cmp(&query_n.score_by(dummy_similarity));
+
+            assert!(ordering == std::cmp::Ordering::Equal || ordering == std::cmp::Ordering::Greater);
```

Comment:
Or are you doing this on purpose to make it more explicit?

```suggestion
            assert_ne!(ordering, std::cmp::Ordering::Less);
```

Reply:
Yours looks more concise 👍